### PR TITLE
[WIP] Use `Intl` API in PluralResolver

### PR DESCRIPTION
**Pull request by Elindorath**
I open this PR in a very early stage to have a place where we can discuss if/how i18next should rely on the `Intl` API to handle plural forms. The goal is to solve #1647.

#### Checklist

- [ ] only relevant code is changed (make a diff before you submit the PR)
- [ ] run tests `npm run test`
- [ ] tests are included

#### Checklist (for documentation change)

- [ ] only relevant documentation part is changed (make a diff before you submit the PR)
- [ ] motivation/reason is provided

### CODE REVIEW

**Comment by adrai** [comment]
fyi: this was already spiked for a major change: https://github.com/adrai/i18next-next
with ordinal and cardinal support

**Comment by Elindorath** [comment]
> fyi: this was already spiked for a major change: [adrai/i18next-next](https://github.com/adrai/i18next-next)
> with ordinal and cardinal support

@adrai Great news!
Is there anything I can help you with?
Would there be a place where we could see the progress?

**Comment by adrai** [comment]
> > fyi: this was already spiked for a major change: [adrai/i18next-next](https://github.com/adrai/i18next-next)
> > with ordinal and cardinal support
> 
> @adrai Great news!
> Is there anything I can help you with?
> Is there already a work in progress on the [i18next-parser](https://github.com/i18next/i18next-parser)?
> Would there be a place where we could see the progress?

as said, this is just a playground for a possible i18next rewrite....

I suggest to continue with your approach by introducing a new v4 format and making a major i18next version for now...

**Comment by adrai** [comment]
ok, seems we will plan for a v21.0.0 that introduces a new default v4 json format (this PR)...
@Elindorath please give us a couple of days to coordinate this... 🙏

**Comment by adrai** [comment]
A question that comes to my mind is: "should we continue to support the simple form? (_plural)"

File: src/PluralResolver.js [hunk 1]
```diff
@@ -112,48 +126,62 @@ class PluralResolver {
...
-        if (suffix === 1) return '';
-        if (typeof suffix === 'number') return `_plural_${suffix.toString()}`;
-        return returnSuffix();
-      } else if (/* v2 */ this.options.compatibilityJSON === 'v2') {
-        return returnSuffix();
-      } else if (/* v3 - gettext index */ this.options.simplifyPluralSuffix && rule.numbers.length === 2 && rule.numbers[0] === 1) {
-        return returnSuffix();
-      }
-      return this.options.prepend && idx.toString() ? this.options.prepend + idx.toString() : idx.toString();
+      return getSuffixRetroCompatible(code, count);
```
Comment (adrai) [inline src/PluralResolver.js:144]: probably this.getSuffixRetroCompatible

**Comment by jamuhl** [comment]
@adrai guess I would prefer one, other for "simple" plurals

File: src/PluralResolver.js [hunk 2]
```diff
@@ -112,48 +126,62 @@ class PluralResolver {
...
-
-      // special treatment for lngs only having singular and plural
-      if (this.options.simplifyPluralSuffix && rule.numbers.length === 2 && rule.numbers[0] === 1) {
-        if (suffix === 2) {
-          suffix = 'plural';
-        } else if (suffix === 1) {
-          suffix = '';
-        }
+      if (this.shouldUseIntlApi()) {
+        return rule.select(count);
```
Comment (adrai) [inline src/PluralResolver.js:141]: probably `return `_${rule.select(count)}`;`

**Comment by Elindorath** [comment]
I am working on the tests, and noticed that some locales currently supported by i18next are not supported by the `Intl` API. Perhaps of more concern, the supported locales are implementation dependent, so may vary from browser to browser.
I couldn't find a proper list so I ran this:
```js
[
  'ach','ak','am','arn','br','fil','gun','ln','mfe','mg','mi','oc','pt','pt-BR','tg','tl','ti','tr','uz','wa','af','an','ast','az','bg','bn','ca','da','de','dev','el','en','eo','es','et','eu','fi','fo','fur','fy','gl','gu','ha','hi','hu','hy','ia','it','kk','kn','ku','lb','mai','ml','mn','mr','nah','nap','nb','ne','nl','nn','no','nso','pa','pap','pms','ps','pt-PT','rm','sco','se','si','so','son','sq','sv','sw','ta','te','tk','ur','yo','ay','bo','cgg','fa','ht','id','ja','jbo','ka','km','ko','ky','lo','ms','sah','su','th','tt','ug','vi','wo','zh','be','bs','cnr','dz','hr','ru','sr','uk','ar','cs','sk','csb','pl','cy','fr','ga','gd','is','jv','kw','lt','lv','mk','mnk','mt','or','ro','sl','he','iw'
].reduce((output, locale, i, locales) => {
  const options = new Intl.PluralRules(locale).resolvedOptions();
  const isSupported = locale.startsWith(options.locale); // Chrome resolve `pt-BR` to `pt`, Firefox keeps it untouched

  return {
    locales: {
      ...output.locales,
      [locale]: isSupported,
    },
    supportedCount: output.supportedCount + (isSupported ? 1 : 0),
    total: locales.length,
  };
}, { locales: {}, supportedCount: 0 })
```
This gives me 114/135 supported locales on Chrome v92.0.4515.131 and 112/135 on Firefox v89.0.2.

Here are the 21 unsupported locales on Chrome:
ach, arn, ay, cnr, csb, dev, gun, ht, iw, mai, mfe, mi, mnk, nap, oc, pms, sco, son, tg, tl, tt

And here are the 23 unsupported locales on Firefox:
ach, an, arn, ay, cnr, csb, dev, gun, ht, iw, jbo, mnk, nah, nap, no, nso, oc, pap, pms, sco, son, tl, wa

Not sure how to address this.
On one hand, the `Intl` API is a supposedly sufficient standard for most people. On the other hand, if the ambition of i18next is to be the most complete i18n framework possible, the `Intl` API becomes a limitation that should be completed. Polyfills like [this one](https://github.com/formatjs/formatjs/tree/main/packages/intl-pluralrules) resolves the issue but the responsibility is left to the i18next user

**Comment by jamuhl** [comment]
good catch...even more as the Intl implementations differ from browser to browser - wasn't aware of that. Overall I agree the Intl API should be sufficient enough and we shouldn't worry too much about loosing some languages on the way - users still can use the older v3 in that case.

**Comment by adrai** [comment]
@Elindorath maybe there is the possibility to use the addRule functionality in case a polyfill is not possible?

**Pull request closed without merging**